server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -669,56 +683,178 @@ public String toString() {
+        private final Semaphore semaphore = new Semaphore(0);
+        private final Iterator<StoreFileMetaData> remainingFiles;
+        private StoreFileMetaData currentFile;
+        private InputStreamIndexInput currentInput = null;
+        private long currentChunkPosition = 0;
+        private final Deque<byte[]> recycledBuffers = ConcurrentCollections.newDeque();
+        private final AtomicReference<Tuple<StoreFileMetaData, Exception>> error = new AtomicReference<>();
+
+        MultiFileSender(Store store, IntSupplier translogOps, StoreFileMetaData[] files, ActionListener<Void> listener) {
+            super(ActionListener.notifyOnce(listener));
+            this.store = store;
+            this.translogOps = translogOps;
+            this.remainingFiles = Arrays.asList(files).iterator();
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            assert ThreadPool.assertCurrentMethodIsNotCalledRecursively();
+            assert Transports.assertNotTransportThread(RecoverySourceHandler.this + "[send file chunk]");
+            while (true) {
+                assert semaphore.availablePermits() == 0;
+                cancellableThreads.checkForCancel();
+                if (canSendMore() == false) {
+                    semaphore.release();
+                    // Here we have to retry before abort to avoid a race situation where the other threads have flipped `canSendMore`
+                    // condition but they are not going to resume the sending process because this thread still holds the semaphore.
+                    final boolean changed = canSendMore() || error.get() != null;
+                    if (changed == false || semaphore.tryAcquire() == false) {
                         break;
                     }
-                    final long requestFilePosition = position;
-                    cancellableThreads.executeIO(() ->
-                        recoveryTarget.writeFileChunk(md, requestFilePosition, content, lastChunk, translogOps.getAsInt(),
-                            ActionListener.wrap(
-                                r -> requestSeqIdTracker.markSeqNoAsProcessed(requestSeqId),
-                                e -> {
-                                    error.compareAndSet(null, Tuple.tuple(md, e));
-                                    requestSeqIdTracker.markSeqNoAsProcessed(requestSeqId);
+                }
+                if (error.get() != null) {
+                    handleErrorOnSendFiles(store, error.get().v2(), new StoreFileMetaData[]{error.get().v1()});
+                    throw error.get().v2();
+                }
+                final FileChunk chunk = readNextChunk();
+                if (chunk == null) {
+                    semaphore.release(); // allow other threads respond if we are not done yet.
+                    if (requestSeqIdTracker.getMaxSeqNo() == requestSeqIdTracker.getProcessedCheckpoint() && semaphore.tryAcquire()) {
```

Comment:
This could also race against a request that is sent and completed between reading getMaxSeqNo and getProcessedCheckpoint. This could mean the numbers turn out equal when we are not finished. This could be solved by reversing the read order here, ie. read out getProcessedCheckpoint first and then getMaxSeqNo.

I would prefer to just see this go single-threaded all of it instead, I find that easier to reason about.